src/System.Memory/src/System/Buffers/Text/Base64Decoder.cs

```diff
@@ -248,52 +292,170 @@ public static OperationStatus DecodeFromUtf8InPlace(Span<byte> buffer, out int b
 
                 if (i0 < 0)
                     goto InvalidExit;
-                WriteThreeLowOrderBytes(ref Unsafe.Add(ref bufferBytes, destIndex), i0);
+
+                WriteThreeLowOrderBytes(ref Unsafe.Add(ref bufferBytes, (IntPtr)destIndex), i0);
                 destIndex += 3;
             }
-            else if (i2 != EncodingPad)
+            else if (t2 != EncodingPad)
             {
-                i2 = Unsafe.Add(ref decodingMap, i2);
+                int i2 = Unsafe.Add(ref decodingMap, (IntPtr)t2);
 
                 i2 <<= 6;
 
                 i0 |= i2;
 
                 if (i0 < 0)
                     goto InvalidExit;
-                Unsafe.Add(ref bufferBytes, destIndex) = (byte)(i0 >> 16);
-                Unsafe.Add(ref bufferBytes, destIndex + 1) = (byte)(i0 >> 8);
+
+                Unsafe.Add(ref bufferBytes, (IntPtr)destIndex) = (byte)(i0 >> 16);
+                Unsafe.Add(ref bufferBytes, (IntPtr)(destIndex + 1)) = (byte)(i0 >> 8);
                 destIndex += 2;
             }
             else
             {
                 if (i0 < 0)
                     goto InvalidExit;
-                Unsafe.Add(ref bufferBytes, destIndex) = (byte)(i0 >> 16);
+
+                Unsafe.Add(ref bufferBytes, (IntPtr)destIndex) = (byte)(i0 >> 16);
                 destIndex += 1;
             }
 
         DoneExit:
-            bytesWritten = destIndex;
+            bytesWritten = (int)destIndex;
             return OperationStatus.Done;
 
         InvalidExit:
-            bytesWritten = destIndex;
+            bytesWritten = (int)destIndex;
             return OperationStatus.InvalidData;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void Avx2Decode(ref byte src, ref byte destBytes, int sourceLength, int destLength, ref uint sourceIndex, ref uint destIndex)
+        {
+            ref byte srcStart = ref src;
+            ref byte destStart = ref destBytes;
+            ref byte simdSrcEnd = ref Unsafe.Add(ref src, (IntPtr)((uint)sourceLength - 45 + 1));
```

Comment:
When calling `Unsafe.Add`, it's legal to create a ref that points to _just past_ the end of the buffer. For example, assume I have a `Span<byte>` _span_ of length 5. Consider the following.

```cs
ref byte a = ref MemoryMarshal.GetReference(span); // &span[0], valid ref
ref byte b = ref Unsafe.Add(ref a, 5); // &span[5], valid ref
ref byte c = ref Unsafe.Add(ref b, 1); // &span[6], *invalid* ref
```

In the above example, both `a` and `b` are valid GC-tracked refs. Since `b` points to memory outside the buffer, it _must not_ be dereferenced. But since `b` points _just beyond_ the end of the buffer, the GC can still track it, so operations like `Unsafe.IsAddressLessThan(a, b)` will still work as expected.

`c`, on the other hand, is further than _just beyond_ the end of the buffer, so the GC cannot track it. If the underlying object moves in memory, the GC is guaranteed to keep `a` and `b` in sync, but it makes no such guarantees for `c`. Therefore comparing `a` and `c` (or `b` and `c`) against each other results in undefined behavior.

The reason I mention this is that we need to be careful when we're creating refs that might be beyond the bounds of the buffer. I don't know from this particular call site if the call is valid from a GC-tracking perspective, so I wanted to draw your attention to it so that you can verify it.

Reply:
(This same comment applies to other instances of `Unsafe.Add`.)

Reply:
Perfect explanation -- thank you!

It is always within the bounds of the buffer. 
At this case here `sourceLength` is guaranteed to be `>= 45`, so `simdSrcEnd` is within the buffer.
The stride is `32`, so it is also within the buffer.

The other places have the same guarantees, as for encoding the min length is >= register size and the stride is register size / 4 * 3, so less than the register size. 
For decoding the stride is the register size, the min length is register size + max two padding + zeros that are written (see comments [here](https://github.com/aklomp/base64/blob/a27c565d1b6c676beaf297fe503c4518185666f7/lib/arch/ssse3/dec_loop.c#L1-L5) and [here](https://github.com/aklomp/base64/blob/a27c565d1b6c676beaf297fe503c4518185666f7/lib/arch/avx2/dec_loop.c#L1-L5), so 24 (SSSE3) or 45 (AVX2).